libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
+	if ndots, exists := rc.Option("ndots"); exists {
+		if n, err := strconv.Atoi(ndots); err != nil || n < 0 {
+			return errdefs.InvalidParameter(
+				fmt.Errorf("invalid number for ndots option: %v", ndots))
+		}
+	}
+	// For each option required by the nameserver, add it if not already
+	// present (if the option already has a value don't change it).
+	for _, opt := range reqdOptions {
+		optName, _, _ := strings.Cut(opt, ":")
+		if _, exists := rc.Option(optName); !exists {
+			rc.AddOption(opt)
+		}
+	}
+
+	rc.md.Transform = "internal resolver"
+	return nil
+}
+
+// Generate returns content suitable for writing to a resolv.conf file. If comments
+// is true, the file will include header information if supplied, and a trailing
+// comment that describes how the file was constructed and lists external resolvers.
+func (rc *ResolvConf) Generate(comments bool) ([]byte, error) {
+	s := struct {
+		Md             *metadata
+		NameServers    []string
+		ExtNameServers []string
+		Search         string
+		Options        string
+		Other          string
+		Overrides      []string
+		Comments       bool
+	}{
+		Md:       &rc.md,
+		Search:   strings.Join(rc.search, " "),
+		Options:  strings.Join(rc.options, " "),
+		Other:    rc.other,
+		Comments: comments,
+	}
+	for _, addr := range rc.nameServers {
+		s.NameServers = append(s.NameServers, addr.String())
+	}
+	for _, ns := range rc.md.ExtNameServers {
+		var nss string
+		if ns.HostLoopback {
+			nss = "host(" + ns.Addr.String() + ")"
+		} else {
+			nss = ns.Addr.String()
+		}
+		s.ExtNameServers = append(s.ExtNameServers, nss)
+	}
+	if rc.md.NSOverride {
+		s.Overrides = append(s.Overrides, "nameservers")
+	}
+	if rc.md.SearchOverride {
+		s.Overrides = append(s.Overrides, "search")
+	}
+	if rc.md.OptionsOverride {
+		s.Overrides = append(s.Overrides, "options")
+	}
+
+	templText := `{{if .Comments}}{{with .Md.HeaderComment}}{{.}}
+
+{{end}}{{end}}{{range .NameServers -}}
+nameserver {{.}}
+{{end}}{{with .Search -}}
+search {{.}}
+{{end}}{{with .Options -}}
+options {{.}}
+{{end}}{{with .Other -}}
+{{.}}
+{{end}}{{if .Comments}}
+# Based on host file: '{{.Md.SourcePath}}'{{with .Md.Transform}} ({{.}}){{end}}
+{{if .Md.UsedDefaultNS -}}
+# Used default nameservers.
+{{end -}}
+{{with .ExtNameServers -}}
+# ExtServers: {{.}}
+{{end -}}
+# Overrides: {{.Overrides}}
+{{with .Md.NDotsFrom -}}
+# Option ndots from: {{.}}
+{{end -}}
+{{end -}}
+`
+	var buf bytes.Buffer
+	templ, err := template.New("summary").Parse(templText)
+	if err != nil {
+		return nil, errdefs.System(err)
+	}
+	if err := templ.Execute(&buf, s); err != nil {
+		return nil, errdefs.System(err)
+	}
+	return buf.Bytes(), nil
+}
+
+// WriteFile generates content and writes it to path. If hashPath is non-zero, it
+// also writes a file containing a hash of the content, to enable UserModified()
+// to determine whether the file has been modified.
+func (rc *ResolvConf) WriteFile(path, hashPath string, perm os.FileMode) error {
```

Comment:
I put it in the package, wrapping Generate(), for a couple of reasons ...

The old exported package functions use it - so, putting it outside the package would have meant duplicating the generate and write.

But, mostly, `sandbox_dns_unix` code depends on not-writing the hash file after it's been set up for the internal nameserver ... after that happens, it's never updated. (If it was, it'd break if the container was later connected to the default network.) That wasn't at all obvious, so I wanted to make it explicit by using a write method that requires a missing hash-file path - rather than just a missing update.

Perhaps the mistake was exporting Generate(). For now, I've un-exported it - but that can change if we want to make the package internal and remove the old methods.

Reply:
> The old exported package functions use it - so, putting it outside the package would have meant duplicating the generate and write.

So? What's the big deal writing a half-dozen lines of similar code in two places?
```go
contents, err := rc.Generate(true)
if err != nil { return err }
if err := somepkg.WriteFile(contents, path, hashPath, 0o644); err != nil {
    return err
}
```

> `sandbox_dns_unix` code depends on

This isn't sandbox_dns_unix code. It is a distinct _exported_ package. Code that is not sandbox_dns_unix [may have no need for writing a digest of a resolv.conf file to disk.](https://en.wikipedia.org/wiki/Interface_segregation_principle) It may not even want to write the generated resolv.conf to disk at all! Maybe it wants to write it into a tarball it's writing, or return it in an RPC response.

I would be much less picky about the shape of the API if this code becomes _exclusively_ an implementation detail of sandbox_dns_unix such that we would be free to make breaking changes at any time to `type ResolvConf`, its methods, or related functions.